addr_manager_ds.go

```diff
@@ -0,0 +1,235 @@
+package peerstore
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	"github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+	mh "github.com/multiformats/go-multihash"
+)
+
+type DatastoreAddrManager struct {
+	ds         ds.Datastore
+	ttlManager *ttlmanager
+	addrSubs   map[peer.ID][]*addrSub
+}
+
+func NewDatastoreAddrManager(ctx context.Context, ds ds.Datastore, ttlInterval time.Duration) *DatastoreAddrManager {
+	mgr := &DatastoreAddrManager{
+		ds:         ds,
+		ttlManager: newTTLManager(ctx, ds, ttlInterval),
+		addrSubs:   make(map[peer.ID][]*addrSub),
+	}
+	return mgr
+}
+
+func (mgr *DatastoreAddrManager) Stop() {
+	mgr.ttlManager.stop()
+}
+
+func peerAddressKey(p *peer.ID, addr *ma.Multiaddr) (ds.Key, error) {
+	hash, err := mh.Sum((*addr).Bytes(), mh.MURMUR3, -1)
+	if err != nil {
+		return ds.Key{}, nil
+	}
+	return ds.NewKey(p.Pretty()).ChildString(hash.B58String()), nil
+}
+
+func (mgr *DatastoreAddrManager) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+func (mgr *DatastoreAddrManager) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	if ttl <= 0 {
+		return
+	}
+
+	mgr.SetAddrs(p, addrs, ttl)
+}
+
+func (mgr *DatastoreAddrManager) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.SetAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+func (mgr *DatastoreAddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	var keys []ds.Key
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+
+		key, err := peerAddressKey(&p, &addr)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		keys = append(keys, key)
+
+		if ttl <= 0 {
+			mgr.ds.Delete(key)
```

Comment:
If you add a `continue` here, the following `else` clause is no longer required.